test/functional/test_framework/address.py

```diff
@@ -41,11 +44,34 @@ def byte_to_base58(b, version):
         str = str[2:]
     return result
 
-# TODO: def base58_decode
+
+def base58_to_byte(s):
+    if not s:
+        return b''
+    n = 0
+    for c in s:
+        n *= 58
+        assert c in chars
+        digit = chars.index(c)
+        n += digit
+    h = '%x' % n
+    if len(h) % 2:
+        h = '0' + h
+    res = binascii.unhexlify(h.encode('utf8'))
+    pad = 0
+    for c in s[:-1]:
+        if c == chars[0]:
+            pad += 1
+        else:
+            break
+    return b'\x00' * pad + res[1:21]
```

Comment:
In commit "tests: implement base58_decode" (9942e0211c7dca24c85a6efcd1de2bd71b95e34d)

I think 21 should be replaced with -4 here to strip off the checksum bytes without assuming a fixed length string. Also it would be good to verify the checksum, maybe optionally with a `verify_checksum=True` default parameter.